### PR TITLE
test(reminders): harden lifecycle synchronization

### PR DESCRIPTION
**Pull request by ReubenBond**
## Problem

Reminder provider suites rely on several independently composed diagnostic waits and test-only service hooks. Their synchronous per-test cleanup is tied to test cancellation, can leave local reminder state behind, and activation-local counter waits can strand a clock driver when an activation is replaced. Lifecycle edge cases around partial startup and failed or superseded range reads also lack focused regression coverage.

## Solution

Introduce a cohesive reminder lifecycle harness for service readiness, topology reconciliation, schedule arming, activation-independent tick completion, sequential refresh, and global quiescence. Convert the shared reminder suite base to asynchronous disposal using an independent bounded cleanup token, clear storage through the grain proxy, refresh every active service in order, and always dispose diagnostics. Preserve the in-memory and SQL Server pre-clear behavior and route Azure Table startup synchronization through the same harness.

Add controlled regressions for activation replacement, current and obsolete reconciliation failures, stalled partial startup cleanup, and cleanup after test cancellation.

## Rationale

Centralizing phase barriers makes provider tests express lifecycle intent without rebuilding synchronization from low-level hooks. Cleanup now establishes both persistent and local-state isolation even when the test token is canceled. Tick completion uses the activation-independent diagnostic event emitted after reminder delivery, matching the approach in #10855 without copying its test-grain cleanup; whichever PR merges second can resolve the small shared-base overlap while retaining this harness.
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/orleans/pull/10891)

### CODE REVIEW

File: test/Orleans.Reminders.Tests/TimerTests/ReminderTestsBase.cs [hunk 1]
```diff
@@ -67,21 +66,70 @@ public ReminderTestsBase(ReminderTestClock reminderClock, InProcessTestCluster h
...
+        Func<CancellationToken, Task> clearReminderTable,
+        Func<CancellationToken, Task> refreshReminderServices,
+        Func<CancellationToken, Task> waitForGlobalQuiescence,
+        TimeSpan timeout)
+    {
+        using var cleanupCancellation = new CancellationTokenSource(timeout);
+        List<Exception>? exceptions = null;
+        await RunPhase(clearReminderTable);
+        await RunPhase(refreshReminderServices);
+        await RunPhase(waitForGlobalQuiescence);
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply test/Orleans.Reminders.Tests/TimerTests/ReminderTestsBase.cs]: Cleanup now stops starting additional phases once its independent budget expires. Added a deterministic regression which verifies later phases are not invoked after expiration.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: test/Orleans.Reminders.Tests/TimerTests/ReminderLifecycleHarness.cs [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+            return (await Task.WhenAll(startedTasks)).Select(started => started.SiloAddress!).ToArray();
+        }
+        catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested)
+        {
+            var missing = siloArray
+                .Where((_, index) => !startedTasks[index].IsCompletedSuccessfully)
+                .Select(silo => silo.SiloAddress);
+            throw new TimeoutException(
+                $"Reminder services did not become ready. Missing silos: {string.Join(", ", missing)}.");
+        }
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply test/Orleans.Reminders.Tests/TimerTests/ReminderLifecycleHarness.cs]: Readiness and topology reconciliation now use separate dedicated timeout and external cancellation tokens. Only dedicated timeout expiration is translated; xUnit/upstream cancellation remains an OperationCanceledException. Added deterministic coverage for both paths.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: test/Orleans.Reminders.Tests/TimerTests/ReminderTestsBase.cs [hunk 3]
```diff
@@ -42,7 +41,7 @@ public class ReminderTestsBase : OrleansTestingBase, IDisposable
     protected const long failAfter = 2;
     protected const long failCheckAfter = 6;
     protected ILogger log;
-    protected ReminderDiagnosticObserver observer;
+    protected ReminderLifecycleHarness observer { get; }
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply test/Orleans.Reminders.Tests/TimerTests/ReminderTestsBase.cs]: Changed this back to a readonly protected field, preserving the existing field naming convention.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

**Comment by ReubenBond** [comment]
Current CI is blocked by a deterministic branch build error: 	est/Orleans.Runtime.Tests/CallbackDataTests.cs(296,37) triggers xUnit1051 because a method accepting CancellationToken is called without TestContext.Current.CancellationToken. This fails the build across the matrix before tests run. Failed run: https://github.com/dotnet/orleans/actions/runs/33143148156
